### ISSUE TITLE
runtime: race between stack shrinking and channel send/recv leads to bad sudog values

### ISSUE THREAD

**Issue by mknyszek** [golang/go#40641]
Internally we've seen a rare crash arise in the runtime since Go 1.14. The error message is typically `sudog with non-nil elem` stemming from a call to `releaseSudog` from `chansend` or `chanrecv`.

The issue here is a race between a mark worker and a channel operation. Consider the following sequence of events. GW is a worker G. GS is a G trying to send on a channel. GR is a G trying to receive on that same channel.

1. GW wants to suspend GS to scan its stack. It calls `suspendG`.
1. GS is about to `gopark` in `chansend`. It calls into `gopark`, and changes its status to `_Gwaiting` BEFORE calling its `unlockf`, which sets `gp.activeStackChans`.
1. GW observes `_Gwaiting` and returns from `suspendG`. It continues into `scanstack` where it checks if it's safe to shrink the stack. In this case, it's fine. So, it reads `gp.activeStackChans`, and sees it as *false*. It begins adjusting `sudog` pointers without synchronization. It reads the `sudog`'s `elem` pointer from the `chansend`, but has not written it back yet.
1. GS continues on its merry way and sets `gp.activeStackChans` and parks. It doesn't really matter when this happens at this point.
1. GR comes in and wants to `chanrecv` on channel. It grabs the channel lock, reads from the `sudog`'s `elem` field, and clears it. GR readies GS.
1. GW then writes the updated sudog's elem pointer and continues on its merry way.
1. Sometime later, GS wakes up because it was readied by GR, and tries to release the `sudog`, which has a non-nil `elem` field.

The fix here, I believe, is to set `gp.activeStackChans` before the `unlockf` is called. Doing this ensures that the value is updated before any worker that could shrink GS's stack observes a useful G status in `suspendG`. This could alternatively be fixed by changing the G status after `unlockf` is called, but I worry that will break a lot of things.

CC @aclements @prattmic

**Comment by mknyszek** [comment]
This should be fixed for Go 1.14 and Go 1.15. It's a bug that was introduced in Go 1.14, and may cause random and unavoidable crashes at any point in time. There may not be enough time to fix this for 1.15 (the failure is very rare, but we've seen it internally), and if not, it should definitely go in a point release.

@gopherbot please open a backport issue for 1.14.

**Comment by mknyszek** [comment]
Oh no! That fix doesn't work because if there's a stack growth at any point in between, we could self-deadlock. Ugh. OK I don't know what the right fix is yet...

**Comment by cherrymui** [comment]
Could we set GS as not safe for shrink stack?

**Comment by mknyszek** [comment]
Yeah, that might work. I'm not sure if it's technically safe for stack scanning though because if it's a `chanrecv` falling into this path then there could be a write into the stack in that same window from a sender? I guess that's already true (or we handle this case) today?

**Comment by cherrymui** [comment]
I think stack scanning should be fine, as it doesn't write to the stack. It is okay to read the old or the new value. As long as the write has a write barrier, it should be fine.

**Comment by mknyszek** [comment]
Thanks Cherry, that makes sense. `sendDirect` has an explicit write barrier and we're not worried about reads of a partially-written pointer in stack scanning because `memmove` guarantees it'll copy a pointer-word a time.

Your solution would mean that we could probably get rid of `syncadjustsudogs` altogether which would be a nice simplification! 😀 A stack shrink will still be enqueued for the next sync preemption so it's probably also fine from a memory use perspective.

**Comment by prattmic** [comment]
If we set it as not safe for stack shrinking, that means we can't shrink stacks of anything blocked on chan send, right? Would that be too limiting, given that such G's may be blocked indefinitely. I certainly regularly see thousands of G's blocked in chan recv, though I imagine chan send is less common.

**Comment by prattmic** [comment]
e.g., in one of the applications where we saw this crash, 44/557 G's were blocked in chan receive, most for >1000 minutes. That is a decent amount of stack space you'll effectively never be able to reclaim.

**Comment by mknyszek** [comment]
There was a state of the world where all this worked, prior to https://golang.org/cl/172982. For 1.14 it's probably safest to revert that CL, since I think it can be cleanly reverted.

For 1.15 (if not backporting) and 1.16 though, @prattmic and I agree would be nice to retain the explicit tracking of when we need to be careful in stack copying that https://golang.org/cl/172982 added, but I'm not sure how to do it safely. It may require some surgery in `gopark` which is a little scary.

@cherrymui what do you think?

**Comment by mknyszek** [comment]
OK actually I think I'm abandoning the idea of the revert. Enough changed after that CL that it doesn't make sense, and after I read some of the old comments, it feels like going back to that state would be subtle.

**Comment by cherrymui** [comment]
Yeah, that CL's description says that it will break the rule when stack shrink can happen synchronously (as it is now).

**Comment by cherrymui** [comment]
Re stack space, for blocked goroutines, we cannot release the used part of the stack regardless. The question is how much stack space we can actually shrink.

**Comment by cherrymui** [comment]
I think it is only unsafe for stack shrinking between setting the G to Gwaiting and the G actually parks, due to the race of `gp.activeStackChans`? Could we block stack shrinking only for that small window? Then in most cases blocked goroutine can still have its stack shrunk.

**Comment by ianlancetaylor** [comment]
That idea makes sense to me.  Block stack shrinking before calling `gopark`, and permit stack shrinking in `chanparkcommit`.

**Comment by ianlancetaylor** [comment]
By the way, I look forward to the test case.

**Comment by mknyszek** [comment]
Thanks @cherrymui, that sounds right to me.

> By the way, I look forward to the test case.

Yeah, that's going to be fun...

**Comment by mknyszek** [comment]
OK I think I have a fix (trybots look good so far, 🤞), but still no test. The goal of the test so far is to create a situation where a goroutine whose stack is ripe for shrinking keeps going to sleep and waking up on channels. Meanwhile, a GC is triggered to try and get a mark worker to race with that goroutine going to sleep. I wrote a test which I've confirmed races stack shrinks with channel sends and receives, but I'm noticing a couple problems in trying to reproduce it. 

1. It's relatively expensive to generate a stack shrink in the runtime, so stress testing this isn't extremely effective. I could maybe get around this by writing an unsafe function in `export_test.go` which does `suspendG` followed by `shrinkstack` in a loop. This way we don't need to wait for the whole GC to end before trying again.
1. The race window is really, really small compared to everything around it. This fact is evidenced not only by looking at the code, but also by the fact that this crash was fairly rare.

As a result of these problems, I'm not certain that this is feasible to write a test for, at least not a short-running one, if we want it to fail reliably. But, maybe I'm just going about this the wrong way (i.e. a stress test is just the wrong choice here)? Any advice?
